x/feeshare/ante/ante.go

```diff
@@ -0,0 +1,117 @@
+package ante
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+
+	feeshare "github.com/classic-terra/core/x/feeshare/types"
+	wasmtypes "github.com/classic-terra/core/x/wasm/types"	
+)
+
+// FeeSharePayoutDecorator Run his after we already deduct the fee from the account with
+// the ante.NewDeductFeeDecorator() decorator. We pull funds from the FeeCollector ModuleAccount
+type FeeSharePayoutDecorator struct {
+	bankKeeper     BankKeeper
+	feesharekeeper FeeShareKeeper
+}
+
+func NewFeeSharePayoutDecorator(bk BankKeeper, fs FeeShareKeeper) FeeSharePayoutDecorator {
+	return FeeSharePayoutDecorator{
+		bankKeeper:     bk,
+		feesharekeeper: fs,
+	}
+}
+
+func (fsd FeeSharePayoutDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	err = FeeSharePayout(ctx, fsd.bankKeeper, feeTx.GetFee(), fsd.feesharekeeper, tx.GetMsgs())
+	if err != nil {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, err.Error())
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// FeePayLogic takes the total fees and splits them based on the governance params
+// and the number of contracts we are executing on.
+// This returns the amount of fees each contract developer should get.
+// tested in ante_test.go
+func FeePayLogic(fees sdk.Coins, govPercent sdk.Dec, numPairs int) sdk.Coins {
+	var splitFees sdk.Coins
+	for _, c := range fees.Sort() {
+		rewardAmount := govPercent.MulInt(c.Amount).QuoInt64(int64(numPairs)).RoundInt()
+		if !rewardAmount.IsZero() {
+			splitFees = splitFees.Add(sdk.NewCoin(c.Denom, rewardAmount))
+		}
+	}
+	return splitFees
+}
+
+// FeeSharePayout takes the total fees and redistributes 50% (or param set) to the contract developers
+// provided they opted-in to payments.
+func FeeSharePayout(ctx sdk.Context, bk BankKeeper, totalFees sdk.Coins, fsk FeeShareKeeper, msgs []sdk.Msg) error {
```

Comment:
can you create an unit test for FeeSharePayout? it seems to be an important piece that extract information from MsgExecuteContract but is missing unit test.

Reply:
Implemented simple test of FeeSharePayoutDecorator